scripts/book_winners.cpp

```diff
@@ -0,0 +1,130 @@
+#include <algorithm>
+#include <filesystem>
+#include <format>
+#include <iostream>
+#include <random>
+#include <ranges>
+#include <string>
+#include <vector>
+
+namespace fs = std::filesystem;
+
+struct Question {
+    int id;
+    std::string porter;
+};
+
+std::string exec(const std::string &cmd);
+std::vector<Question> get_porters(const fs::path &questions_dir);
+std::optional<std::string> get_porter(const fs::path &question);
+std::vector<Question> get_winners(const std::vector<Question> &questions);
+
+int main() {
+    const auto cwd = fs::current_path();
+    const auto questions_dir = cwd / "questions";
+    if (!fs::exists(questions_dir)) {
+        throw std::runtime_error(
+            "Please run this from the root of the porting repo");
+    }
+
+    const std::vector<Question> ported_questions = get_porters(questions_dir);
+    std::cout << std::format("Found {} questions ported by someone else\n",
+                             ported_questions.size());
+    std::vector<Question> winners = get_winners(ported_questions);
+    std::cout << "\n--- WINNERS ---\n";
+    for (const auto &winner : winners) {
+        std::cout << std::format("{} (ported question {})\n", winner.porter,
+                                 winner.id);
+    }
+}
+
+std::string trim_trailing_newline(std::string str) {
+    str.erase(std::find_if(str.rbegin(), str.rend(),
+                           [](char ch) { return ch != '\n'; })
+                  .base(),
+              str.end());
+    return str;
+}
+
+std::vector<Question> get_porters(const fs::path &questions_dir) {
+    std::vector<Question> ported_questions;
+    for (const auto &question_dir : fs::directory_iterator{questions_dir}) {
+        const auto porter =
+            get_porter(fs::path{"questions"} / question_dir.path().filename());
+        if (!porter.has_value()) {
+            continue;
+        }
+        const int id = std::stoi(question_dir.path().filename());
+        ported_questions.push_back(Question{id, *porter});
+    }
+    return ported_questions;
+}
+
+// Update these shas next time we port the questions
+std::optional<std::string> get_porter(const fs::path &question) {
+    const std::string last_commit_before_porting{
+        "6349ecaf7e910e22ca0c94c197906b750146f821"};
+    // These just fix formatting or in other ways mask the commit that did the
+    // bulk of the work
```

Comment:
Maybe we can use some heuristic to detect such commits? For example, taking only the first commit that touched the question or looking for the string "Fixes ..." in the commit message. (I want to go over the questions once more to check the formatting according to our last observations, so it's possible that more commits will end up in this vector.)

Reply:
I don't think it's worth the effort. Sometimes it's less work to just do some manual work than to try to automate it, and I think this is one of those cases. Even if we think of and implement a heuristic, it's probably about the same work to check that the heuristic is correct as it is to simply update this list. (I'm planning to make a final pass when the porting is complete.)